### ISSUE TITLE
Unable to use aws_credentials_file as in S3 output

### ISSUE THREAD

**Issue by tkocheran-chromeriver** [logstash-plugins/logstash-input-s3#28]
I'm trying to use this S3 input plugin to read files previously written by the S3 output plugin. 

The following output config works just fine:

```
input { 
    file { 
        path => [ '/var/log/syslog' ]
    }
}

output {
    s3 {
        aws_credentials_file => '/srv/logstash/config/input.conf'
        bucket => 'mahbucket'
        use_ssl => true
    }
}
```

However, the same configuration for using S3 input fails:

```
input {
    s3 {
        aws_credentials_file => '/srv/logstash/config/input.conf'
        bucket => 'mahbucket'
        use_ssl => true
    }
}

output {
    stdout { }
}
```

I get the following errors:

```
A plugin had an unrecoverable error. Will restart this plugin.
  Plugin: <LogStash::Inputs::S3 aws_credentials_file=>"/srv/logstash/config/aws-credentials.yml", bucket=>"mahbucket", region=>"us-east-1", temporary_directory=>"/tmp/logstash">
  Error: 
Missing Credentials.

Unable to find AWS credentials.  You can configure your AWS credentials
a few different ways:

* Call AWS.config with :access_key_id and :secret_access_key

* Export AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY to ENV

* On EC2 you can run instances with an IAM instance profile and credentials
  will be auto loaded from the instance metadata service on those
  instances.

* Call AWS.config with :credential_provider.  A credential provider should
  either include AWS::Core::CredentialProviders::Provider or respond to
  the same public methods.

= Ruby on Rails

In a Ruby on Rails application you may also specify your credentials in
the following ways:

* Via a config initializer script using any of the methods mentioned above
  (e.g. RAILS_ROOT/config/initializers/aws-sdk.rb).

* Via a yaml configuration file located at RAILS_ROOT/config/aws.yml.
  This file should be formated like the default RAILS_ROOT/config/database.yml
  file.

 {:level=>:error}
```

I'm using the same AWS credentials file for input and for output and it works for output but not for input, hence I'm filing this bug.

I'm using Logstash 1.5.0.rc2 and Logstash S3 Input 0.1.8.

**Comment by jxubc** [comment]
you need just add

require "logstash/plugin_mixins/aws_config"

and clean up the duplicated configs

Regards,
Jie

**Comment by arabold** [comment]
I _think_ this issue can be set to resolved. I was running into the same problem with older Logstash releases but the 1.5.1 distribution worked out of the box.

**Comment by arabold** [comment]
I'll take that back... it seems using `aws_credentials_file` isn't working with 1.5.1.

**Comment by arabold** [comment]
I've spend more time testing. It looks like this works when using IAM roles. I have full access to S3 without providing credentials. However, it does *NOT* work when using an external credentials file like this:

```
  s3 {
    type => "cloudtrail"
    codec => "cloudtrail"
    aws_credentials_file => "/etc/logstash/conf.d/aws_credentials.yml"
    bucket => "xxx"
    prefix => "cloudtrail/"
    region => "us-east-1"
    backup_to_bucket => "xxx"
    backup_add_prefix => "_processed/"
    delete => true
    add_field => {
      "_index_name" => "cloudtrail"
    }
  }
```

If I use this configuration, the following error is thrown:
```
{:timestamp=>"2015-08-07T18:13:54.737000+0000", :message=>"A plugin had an unrecoverable error. Will restart this plugin.\n  Plugin: <LogStash::Inputs::S3 type=>\"cloudtrail\", codec=><LogStash::Codecs::CloudTrail spool_size=>50>, aws_credentials_file=>\"/etc/logstash/conf.d/aws_credentials.yml\", bucket=>\"xxx\", prefix=>\"cloudtrail/\", region=>\"us-east-1\", backup_to_bucket=>\"xxx\", backup_add_prefix=>\"_processed/\", delete=>true, add_field=>{\"_index_name\"=>\"cloudtrail\"}, debug=>false, use_ssl=>true, interval=>60, temporary_directory=>\"/tmp/logstash\">\n  Error: Access Denied", :level=>:error}
```

However, if I add the S3 access to the IAM role of the instance, it starts working immediately. So it looks like +the IAM role takes precedence over the credentials file+.

**Comment by jstangroome** [comment]
The problem lies here:
https://github.com/logstash-plugins/logstash-input-s3/blob/master/lib/logstash/inputs/s3.rb#L353-L361

The `else` block of the if statement is never reached because `@credentials` is always truthy, even when it is an empty array.

**Comment by jstangroome** [comment]
Appears to be fixed in PR #46 awaiting merge.
